ycmd.el

```diff
@@ -352,6 +357,23 @@ Returns the new value of `ycmd-force-semantic-completion`.
              (if force "enabled" "disabled"))
     (setq ycmd-force-semantic-completion force)))
 
+(defun ycmd--get-tags-file (buffer)
+  "Get tags files for current BUFFER or nil.
+
+Use `tags-file-name' variable if set, or look up directory
+hierarchy for first found tags file."
+  (or tags-file-name
+      (let* ((directory (locate-dominating-file
```

Comment:
It seems like we might need a combination of the two. I think the `locate-dominating-file` approach is good for a lot of people, but they might need to search for files named something other than `tags` or 'TAGS`, e.g. if they use `CTAGS` as their tag file name by default. But the `tags-file-name` idea will be useful if people use shared tag files. 

Honestly, I don't know enough about how people use tag files to be positive. I think we should create a system that seems reasonable, and which addresses the needs of a known user. We can polish it as we get input.